### ISSUE TITLE
Exiv2::Image::xmpPacket() does not work for some tiff.

### ISSUE THREAD

**Issue by t246246** [Exiv2/exiv2#226]
I'm using exiv2 on macOS (High Sierra), exiv2 0.26 001a00 (64 bit build), comes from macport.
First I prepare 2 images using ImagaMagick (6.9.9-26 Q16 x86_64), just do the following:
```
convert logo: logo.tif
convert logo: logo.jpg
```
then, add simple xmp with exiv2.
```
exiv2 -M"set Xmp.dc.rights foo" logo.tif 
exiv2 -M"set Xmp.dc.rights foo" logo.jpg
```
Here is the file after adding xmp:
[logo.zip](https://github.com/Exiv2/exiv2/files/1704875/logo.zip)

I confirmed the raw packet is there by using `exiv2 -pX`, so far it looks no problem.

Then I compile [`xmpdump.cpp`](https://github.com/Exiv2/exiv2/blob/0.26/src/xmpdump.cpp) found in your src directory, with gcc, and just execute.
For jpeg image, it outputs xpacket in the file, however for tiff, it returns error:
```Caught Exiv2 exception './logo.tif: No XMP packet found in the file'```
It looks it fails to find xpacket.

I think this should be worked for all supported image, but I am afraid i missed something - 
are there any way to explicitly specify the page, because tiff could have multiple images in one file.

---
PS
I checked with MSYS2 binaries (mingw-w64-x86_64-exiv2-0.26-3) on Windows 7 (64 bit), and the result are the same.

**Comment by clanmills** [comment]
Please attach your file so that I can investigate.  For historical reasons, exiv2 has two ways to locate XMP data in an image.  It's possible (although unusual) for -pX to locate XMP that isn't detected by another command options such as setting XMP values.  The most common likely is multi-page tiff and your comment about explicitly specifying a page causes me to believe that is your use-case.

Exiv2 has never supported multi-page tiff files.  We have an project to deal with this and BigTiff (64 bit tiff).  This is a major change to Exiv2 and we don't have a schedule at this time.

A couple of years ago, I investigated using the libtiff utilities **_tiffsplit_** to break a multi-page tiff into separate files which could be modified with exiv2, and recombined into a single file using **_tiffcp_**.   http://dev.exiv2.org/issues/1124#note-4

**Comment by t246246** [comment]
Thank you for your comment, especially for multi-page tiff.
For now, I'm using just single page tiff.  I've just attatched 2 files in zip format in the first message.

Maybe something is wrong with this tiff, even tiffcp lost xmp information.
```
% tiffcp logo.tif logo.tif logo2.tif
% echo $?
0
% exiv2 -pX logo2.tif
```
outputs nothing.

**Comment by clanmills** [comment]
Thanks for the test files which I didn't notice earlier.

As you have said, this is a single image tiff file (and a jpg).  I don't see anything unusual about either file.  I can run your command to update the metadata.
```
  723 rmills@rmillsmbp:~/Downloads/logo $ exiv2 -pR logo.jpg
STRUCTURE OF JPEG FILE: logo.jpg
 address | marker       |  length | data
       0 | 0xffd8 SOI  
       2 | 0xffe0 APP0  |      16 | JFIF............
      20 | 0xffe1 APP1  |    2492 | http://ns.adobe.com/xap/1.0/.<?x
    2514 | 0xffdb DQT   |      67 
    2583 | 0xffdb DQT   |      67 
    2652 | 0xffc0 SOF0  |      17 
    2671 | 0xffc4 DHT   |      30 
    2703 | 0xffc4 DHT   |      78 
    2783 | 0xffc4 DHT   |      29 
    2814 | 0xffc4 DHT   |      71 
    2887 | 0xffda SOS  
724 rmills@rmillsmbp:~/Downloads/logo $ exiv2 -pR logo.tif 
STRUCTURE OF TIFF FILE (II): logo.tif
 address |    tag                              |      type |    count |    offset | value
      10 | 0x0100 ImageWidth                   |     SHORT |        1 |           | 640
      22 | 0x0101 ImageLength                  |     SHORT |        1 |           | 480
      34 | 0x0102 BitsPerSample                |     SHORT |        1 |           | 8
      46 | 0x0103 Compression                  |     SHORT |        1 |           | 5
      58 | 0x0106 PhotometricInterpretation    |     SHORT |        1 |           | 3
      70 | 0x010a FillOrder                    |     SHORT |        1 |           | 1
      82 | 0x0111 StripOffsets                 |      LONG |        1 |           | 4318
      94 | 0x0112 Orientation                  |     SHORT |        1 |           | 1
     106 | 0x0115 SamplesPerPixel              |     SHORT |        1 |           | 1
     118 | 0x0116 RowsPerStrip                 |     SHORT |        1 |           | 480
     130 | 0x0117 StripByteCounts              |      LONG |        1 |           | 28156
     142 | 0x011c PlanarConfiguration          |     SHORT |        1 |           | 1
     154 | 0x0129 PageNumber                   |     SHORT |        2 |           | 0 1
     166 | 0x013e WhitePoint                   |  RATIONAL |        2 |       218 | 1343036287/4294967295 1413044223/4294967295
     178 | 0x013f PrimaryChromaticities        |  RATIONAL |        6 |       234 | 2748779007/4294967295 1417339263/4294967295 1288490239/4294967295 2576980479/4294967295 644245119/4294967295 ...
     190 | 0x0140 ColorMap                     |     SHORT |      768 |       282 | 1028 1542 2570 5140 4883 ...
     202 | 0x02bc XMLPacket                    |      BYTE |     2499 |      1818 | <?xpacket begin="..." id="W5M0Mp ...
END logo.tif
725 rmills@rmillsmbp:~/Downloads/logo $
```
I believe I have the current macports/exiv2 installed which is v0.26.  Same result as above (which were produced with the current GitHub 'master').  Can you send me bin/exiv2 and lib/libexiv2.26.0.0.dylib from your MacPorts installation (for me, this is in /opt/local/bin and /opt/local/lib).

Tomorrow, I will build and install the latest MacPorts/exiv2 on a machine on which I've never used MacPorts.

--------------

I've looked at your command: `$ tiffcp logo.tif logo.tif logo2.tif`.  I think there's a bug in tiffcp that has failed to copy the XMPPacket to logo2.tif.  You'll have to discuss that with the engineers at libtiff.

```
578 rmills@rmillsmbp:~/Downloads/logo $ tiffcp logo.tif  logo.tif logo2.tif
579 rmills@rmillsmbp:~/Downloads/logo $ exiv2 -pR logo2.tif 
STRUCTURE OF TIFF FILE (II): logo2.tif
 address |    tag                              |      type |    count |    offset | value
   28166 | 0x0100 ImageWidth                   |     SHORT |        1 |           | 640
   28178 | 0x0101 ImageLength                  |     SHORT |        1 |           | 480
   28190 | 0x0102 BitsPerSample                |     SHORT |        1 |           | 8
   28202 | 0x0103 Compression                  |     SHORT |        1 |           | 5
   28214 | 0x0106 PhotometricInterpretation    |     SHORT |        1 |           | 3
   28226 | 0x010a FillOrder                    |     SHORT |        1 |           | 1
   28238 | 0x0111 StripOffsets                 |      LONG |        1 |           | 8
   28250 | 0x0112 Orientation                  |     SHORT |        1 |           | 1
   28262 | 0x0115 SamplesPerPixel              |     SHORT |        1 |           | 1
   28274 | 0x0116 RowsPerStrip                 |     SHORT |        1 |           | 480
   28286 | 0x0117 StripByteCounts              |      LONG |        1 |           | 28156
   28298 | 0x011c PlanarConfiguration          |     SHORT |        1 |           | 1
   28310 | 0x0129 PageNumber                   |     SHORT |        2 |           | 0 0
   28322 | 0x013e WhitePoint                   |  RATIONAL |        2 |     28362 | 335759072/1073741824 353261056/1073741824
   28334 | 0x013f PrimaryChromaticities        |  RATIONAL |        6 |     28378 | 687194752/1073741824 354334816/1073741824 322122560/1073741824 644245120/1073741824 161061280/1073741824 ...
   28346 | 0x0140 ColorMap                     |     SHORT |      768 |     28426 | 1028 1542 2570 5140 4883 ...
   58120 | 0x0100 ImageWidth                   |     SHORT |        1 |           | 640
   58132 | 0x0101 ImageLength                  |     SHORT |        1 |           | 480
   58144 | 0x0102 BitsPerSample                |     SHORT |        1 |           | 8
   58156 | 0x0103 Compression                  |     SHORT |        1 |           | 5
   58168 | 0x0106 PhotometricInterpretation    |     SHORT |        1 |           | 3
   58180 | 0x010a FillOrder                    |     SHORT |        1 |           | 1
   58192 | 0x0111 StripOffsets                 |      LONG |        1 |           | 29962
   58204 | 0x0112 Orientation                  |     SHORT |        1 |           | 1
   58216 | 0x0115 SamplesPerPixel              |     SHORT |        1 |           | 1
   58228 | 0x0116 RowsPerStrip                 |     SHORT |        1 |           | 480
   58240 | 0x0117 StripByteCounts              |      LONG |        1 |           | 28156
   58252 | 0x011c PlanarConfiguration          |     SHORT |        1 |           | 1
   58264 | 0x0129 PageNumber                   |     SHORT |        2 |           | 1 0
   58276 | 0x013e WhitePoint                   |  RATIONAL |        2 |     58316 | 335759072/1073741824 353261056/1073741824
   58288 | 0x013f PrimaryChromaticities        |  RATIONAL |        6 |     58332 | 687194752/1073741824 354334816/1073741824 322122560/1073741824 644245120/1073741824 161061280/1073741824 ...
   58300 | 0x0140 ColorMap                     |     SHORT |      768 |     58380 | 1028 1542 2570 5140 4883 ...
END logo2.tif
580 rmills@rmillsmbp:~/Downloads/logo $ 
```

**Comment by t246246** [comment]
Here is my binary you requested.
[exiv2.zip](https://github.com/Exiv2/exiv2/files/1705028/exiv2.zip)

I will try with git master later.

**Comment by clanmills** [comment]
Thanks for sending this.  I'm in bed now as it's 23:00/11pm in England.  I'll look at this tomorrow.

**Comment by clanmills** [comment]
I've understood your issue.  You can extract the "raw" XMP with the following API:

```
image->printStructure(std::cout,Exiv2::kpsXMP,0);
```
I've modified src/xmpdump.cpp as follows:
```
// ***************************************************************** -*- C++ -*-
// xmpdump.cpp
// Sample program to dump the XMP packet of an image

#include "image.hpp"
#include <string>
#include <iostream>
#include <cassert>

int main(int argc, char* const argv[])
try {

    if (argc != 2) {
        std::cout << "Usage: " << argv[0] << " file\n";
        return 1;
    }

    Exiv2::Image::AutoPtr image = Exiv2::ImageFactory::open(argv[1]);
    assert(image.get() != 0);
    image->printStructure(std::cout,Exiv2::kpsXMP,0);

    return 0;
}
catch (Exiv2::AnyError& e) {
    std::cout << "Caught Exiv2 exception '" << e << "'\n";
    return -1;
}
```

I can successfully build and run this as follows:
```
529 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/src $ clang++ xmpdump.cpp -o xmpdump -I ../include/exiv2 -I ../build -lexiv2
530 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/src $ ./xmpdump ~/Downloads/logo/logo.tif  | xmllint --format -
<?xml version="1.0"?>
<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 4.4.0-Exiv2">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description xmlns:dc="http://purl.org/dc/elements/1.1/" rdf:about="">
      <dc:rights>
        <rdf:Alt>
          <rdf:li xml:lang="x-default">foo man choo is alive in england</rdf:li>
        </rdf:Alt>
      </dc:rights>
    </rdf:Description>
  </rdf:RDF>
</x:xmpmeta>
<?xpacket end="w"?>
531 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/src $ 
```

**_Discussion_**

1) **_Thanks for sending `bin/exiv2` and `lib/libexiv2.26.dylib`_**
These are the same as the ones I have build with macports.

2) **_Misunderstanding of the issue definition_**
I mistakenly thought that you were unable to execute the command:
```
$ exiv2 -M"set Xmp.dc.rights foo man choo is alive in England" logo.tif
```
I now realise that you have successfully executed that command.  Your issue is that you are unable to recover the "raw" XMP with the API image->xmpPacket()

3) _**src/xmpdump.cpp**_
I've never noticed this program in the src/ directory and I have been working with the Exiv2 project for 10 years!  It should be moved to the samples/ directory and build along with other samples.

4) **_API image->xmpPacket()_**

I don't know why `image->xmpPacket();` is crashing on your file.  That API (and all the APIs) are intended to be independent of the image format (tif/jpeg/png etc).  When I restore the original code in xmpdump.cpp, build and run it on my favourite test image (stored on my website), I see:
```
536 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/src $ ./xmpdump http://clanmills.com/Stonehenge.jpg | xmllint --format -
<?xml version="1.0"?>
<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 4.4.0-Exiv2">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description xmlns:xmp="http://ns.adobe.com/xap/1.0/" xmlns:dc="http://purl.org/dc/elements/1.1/" rdf:about="" xmp:Rating="0" xmp:ModifyDate="2015-07-16T20:25:28+01:00">
      <dc:description>
        <rdf:Alt>
          <rdf:li xml:lang="x-default">Classic View</rdf:li>
        </rdf:Alt>
      </dc:description>
    </rdf:Description>
  </rdf:RDF>
</x:xmpmeta>
<?xpacket end="w"?>
537 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/src $ 
```
5) **_MinGW Build_**

It's clear that there is an issue in `image->xmpPacket()` with tiff files.  You have confirmed that this is present on other platforms.

6) **_The API Image:printStructure()_**

I added this API in v0.25 with kpsBasic and kpsXMP and expanded it in v0.26 with the addition of kpsRecursive and kpsICC.  The exiv2 command-line options `-pS, -pX, -pR, -pC` use that API.  I added this as a debugging feature to enable me to understand the structure of various image formats.  Once developmed, I realised that printStructure() was a useful API with its own merit.  This is how/why Exiv2 v0.25 and v0.26 has 2 image parsers.  There is a major project in progress to work on BigTiff and multi-page tiff and it's likely that a future version of Exiv2 will consolidate the image parsing code.  

7) Difference between image->printStructure() and image->xmpPacket()

Although the result of those APIs is similar, the implementation is very different.  image->printStructure() extracts the XMP from the file.  image->xmpPacket() serialises the XMP in the image object (using Adobe XMPsdk) and returns it.

8) **_Action from here_**

As I have provided the other API for extracting "raw" XMP, you have an effective work-around that you can use with the shipping product.  I therefore don't see this as an urgent matter.

I don't want to get further involved in the resolution of this issue.  I've worked on Exiv2 for 10 years and wished to retire at the end of 2017 (as my 67th birthday gift).  I am delighted to have recruited 3 new contributors whom I will be happy to mentor/support with this issue.  The consequential matters that require attention are:

a) Why does image->xmpPacket() crash on your file?
b) Move src/xmpdump.cpp to samples/xmpdump.cpp and build it
c) Appropriate updates to the test suite

**Comment by t246246** [comment]
I greatly appreciate your kind help. Image:printStructure() is good enough for me, it is rather preferable as that does not depend on Adobe XMPsdk.

**Comment by clanmills** [comment]
If you wish to open the file and read the "raw" XMP, _image->printStructure()_ is the most effective way to achieve that goal. 
 
Several users have requested that Adobe XMPsdk be an external library and to not compile/embed Adobe's code into Exiv2.  We have a work-in-progress project to enable Exiv2 to be built using a choice of the existing XMPsdk code in the repository, or to download and build Adobe's code.  The users can choose the edition (2016 or 2014 or 2013).

Adobe XMPsdk is an important **_optional_** sub-system within Exiv2.  The purpose of using Adobe's XMPsdk in Exiv2 is to enable editing of XMP metadata.  When _image->readMetadata()_ is called, the XMP is located in the image and fed into an Adobe XMPsdk object stored in the Exiv2::Image instance.  Commands to edit/delete/insert metadata are executed on the XMPsdk object.  The  user call obtain the "current" XMP by serialising that object using _image->xmpPacket()_.  When you call _image->writeMetadata()_, the XMP is serialised and written to file.

_image->printStructure()_ always extracts the XMP from the image file itself.  It knows nothing about what happens during the work-flow _image->readMetadata() ... edit ... image->writeMetadata()_

Exiv2 has always supported a build option to omit all support for XMP metadata.  In effect, no Adobe  XMPsdk is built into the library.  It's likely that _image->printMetadata()_ extracts "raw" XMP in all versions of Exiv2.  If you wish to edit XMP, you must enable XMPsdk in the build because _image->writeMetadata()_ only writes XMP to the image file when the library has been built to use Adobe XMPsdk.

**Comment by clanmills** [comment]
This issue applies to all tiff files.  It's amazing that this has never been reported!

Restating the actions required (from above):

> 8 **_Action from here_**
> a) Why does image->xmpPacket() return an empty string in your file?
> b) Move src/xmpdump.cpp to samples/xmpdump.cpp and build it
> c) Appropriate updates to the test suite

Here's the fix for the code for tiff files.  In src/tiffimage.cpp about line 190.  Add `setXmpPacket(xmpData_.xmpPacket())` following `setByteOrder(bo)` following `TiffParser::decode(...)`
```
        ByteOrder bo = TiffParser::decode(exifData_,
                                          iptcData_,
                                          xmpData_,
                                          io_->mmap(),
                                          (uint32_t) io_->size());
        setByteOrder(bo);
        setXmpPacket(xmpData_.xmpPacket());  // set the XMP string into this->xmpPacket_
        // read profile from the metadata
```

**Comment by clanmills** [comment]
@piponazo 

Can you put the following change into the code, please.  Then we can close this one.
```
580 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/build $ git diff ../src/image.cpp 
diff --git a/src/image.cpp b/src/image.cpp
index e0eebf3d..a718f43c 100644
--- a/src/image.cpp
+++ b/src/image.cpp
@@ -562,6 +562,12 @@ namespace Exiv2 {
 
     std::string& Image::xmpPacket()
     {
+        // Serialize the current XMP
+        if (xmpData_.count() > 0 && !writeXmpFromPacket()) {
+            XmpParser::encode(xmpPacket_, xmpData_,
+                              XmpParser::useCompactFormat |
+                              XmpParser::omitAllFormatting);
+        }
         return xmpPacket_;
     }
 
581 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/build $ 
```

**Comment by piponazo** [comment]
Sure!

**Comment by clanmills** [comment]
This is the issue that is being addressed by PR #531.  I won't close this at the moment, however I will remove it from Milestone v0.27 as this causes the "open for v0.27" to decrease.  Tomorrow (2018-11-22/Thankgiving) I will focus on #531 which is cause Travis running Ubuntu 14.04/Clang5/ASAN to fail in icc-test.sh

**Comment by clanmills** [comment]
This issue is solved.  Releasing the code is being delayed by #531 which is one of two issues remaining to be solved for Exiv2 v0.27 RC3.  I'm closing this because the issue has been investigated and solved.  Tomorrow I'll spend more time on #531 which is a test failure in test/icc-test.sh on Clang-5 build on Ubuntu-14.04.

**Comment by piponazo** [comment]
I am confused. #531 is the PR that should solve this issue (226) but we still have the issues with the clang-5 CI job. Or you have found a solution already?

In case you did not solve the problem yet, this issue should remain open, until we can merge #531.

**Comment by clanmills** [comment]
I wanted to get the open issue as low as possible.  I'll solve the buildserver puzzle tomorrow.  I'm surprised that I haven't found it already.  I couldn't get conan to install on ubuntu 14.04.  We'll get there.  I'll reopen this, although I don't expect this issue to require work!  It's #531 that's the deal-breaker.

**Comment by clanmills** [comment]
I believe this is fixed with 74a2223d on branch 0.27-RC3.

```
675 rmills@rmillsmm:~/gnu/github/exiv2/exiv2/build $ bin/xmpdump ~/Downloads/logo/logo.tif | xmllint --format -
<?xml version="1.0"?>
<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 4.4.0-Exiv2">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description xmlns:dc="http://purl.org/dc/elements/1.1/" rdf:about="">
      <dc:rights>
        <rdf:Alt>
          <rdf:li xml:lang="x-default">foo</rdf:li>
        </rdf:Alt>
      </dc:rights>
    </rdf:Description>
  </rdf:RDF>
</x:xmpmeta>
<?xpacket end="w"?>
676 rmills@rmillsmm:~/gnu/github/exiv2/exiv2/build $ exiv2 -px ~/Downloads/logo/logo.tif 
Xmp.dc.rights                                LangAlt     1  lang="x-default" foo
677 rmills@rmillsmm:~/gnu/github/exiv2/exiv2/build $
```

**Comment by clanmills** [comment]
I don't know what's going on here.  When you reported this last week, it was Ubuntu 14.04 and the branch exiv2/piponazo branch fix#226.  I was able to reproduce.  However, I could apply/remove the code concerning XMPPacket and reproduce the failures in test/icc-test.sh and test/webp-test.sh  I could demonstrate that the XMPPacket issue is solved by the change in Modify Image::xmpPacket().  There's no evidence that the test harness failures was related to Modify Image::xmpPacket().

Now, we have something similar on 0.27-RC3 with https://travis-ci.org/Exiv2/exiv2/jobs/462279634

I would like you to close this issue and #531 and open a new issue about https://travis-ci.org/Exiv2/exiv2/jobs/462279634  Do you know the version of Linux on which that is failing as I have been unable to reproduce this on Ubuntu 18.04 with Clang 5.0.0.

**Comment by piponazo** [comment]
Fixed by #583
